test/auth_test.go

```diff
@@ -359,3 +360,281 @@ func TestPermViolation(t *testing.T) {
+		}
+		_, ok := resp.Header["X-NATS-Result"]
+		if !ok {
+			t.Error("Missing header in response")
+		}
+
+		// Server injects this header.
+		_, ok = resp.Header["Nats-Request-Info"]
+		if !ok {
+			t.Error("Missing header in response")
+		}
+	})
+
+	t.Run("multiple RespondMsg stream", func(t *testing.T) {
+		_, err := ncA.Subscribe("quux", func(m *nats.Msg) {
+			m.Data = []byte("start")
+			m.Header["Task-Progress"] = []string{"0%"}
+			err = m.RespondMsg(m)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			m.Header["Task-Progress"] = []string{"50%"}
+			m.Data = []byte("wip")
+			err = m.RespondMsg(m)
+			if err != nil {
+				errCh <- err
+				return
+			}
+			m.Header["Task-Progress"] = []string{"100%"}
+			m.Data = []byte("done")
+			err = m.RespondMsg(m)
+			if err != nil {
+				errCh <- err
+				return
+			}
+		})
+		if err != nil {
+			t.Fatal(err)
+		}
+		ncA.Flush()
+
+		inbox := nats.NewInbox()
+		responses, err := ncB.SubscribeSync(inbox)
+		if err != nil {
+			errCh <- err
+			return
+		}
+		ncB.Flush()
+		err = ncB.PublishRequest("quux", inbox, []byte("start"))
+		if err != nil {
+			t.Error(err)
+		}
+
+		getNext := func(t *testing.T, sub *nats.Subscription) *nats.Msg {
+			resp, err := sub.NextMsg(2 * time.Second)
+			if err != nil {
+				t.Fatal(err)
+			}
+			if len(resp.Header) != 2 {
+				t.Errorf("Expected original headers as well, got: %d", len(resp.Header))
+			}
+			return resp
+		}
+
+		resp := getNext(t, responses)
+		got = string(resp.Data)
+		expect = "start"
+		if got != expect {
+			t.Errorf("Expected %v, got: %v", expect, got)
+		}
+		v, ok := resp.Header["Task-Progress"]
+		if !ok {
+			t.Error("Missing header in response")
+		}
+		got = v[0]
+		if got != "0%" {
+			t.Errorf("Unexpected value in header, got: %v", got)
+		}
+
+		resp = getNext(t, responses)
+		got = string(resp.Data)
+		expect = "wip"
+		if got != expect {
+			t.Errorf("Expected %v, got: %v", expect, got)
+		}
+		v, ok = resp.Header["Task-Progress"]
+		if !ok {
+			t.Error("Missing header in response")
+		}
+		got = v[0]
+		if got != "50%" {
+			t.Errorf("Unexpected value in header, got: %v", got)
+		}
+
+		resp = getNext(t, responses)
+		got = string(resp.Data)
+		expect = "done"
+		if got != expect {
+			t.Errorf("Expected %v, got: %v", expect, got)
+		}
+		v, ok = resp.Header["Task-Progress"]
+		if !ok {
+			t.Error("Missing header in response")
+		}
+		got = v[0]
+		if got != "100%" {
+			t.Errorf("Unexpected value in header, got: %v", got)
+		}
+	})
+
+	t.Run("RespondMsg different reply", func(t *testing.T) {
+		inbox := nats.NewInbox()
+		sub, err := ncA.SubscribeSync(inbox)
+		if err != nil {
+			t.Fatal(err)
+		}
+		_, err = ncA.Subscribe("baz", func(m *nats.Msg) {
+			// Publish new message with different reply from to sub in Account A.
+			msg := &nats.Msg{
+				Header: m.Header,
+				Data:   []byte("baz"),
+				Reply:  inbox,
+			}
+			err = m.RespondMsg(msg)
```

Comment:
You get a RACE when running the tests and I think it is because of `err` here. I would use a local variable here.
```
if err := m.RespondMsg(msg); err != nil {
    errCh <- err
}
```

Reply:
ah right, thanks Ivan